Fix content length handling when filtering responses

candrews opened the pull request:
Move setContentLength fitlering to FilterServletResponseWrapper
*    All FilterServletResponseWrapper implemtations need to ignore the
*    content length, not just CompressionServletResponseWrapper.

In FilterServletResponseWrapper, ignore other ways to set ContentLength
*    If the Content-Length is set in any way, filtering will not work
*    correctly; the response will be truncated at the content length.

candrews commented:
Build error is unrelated. That error is addressed by https://github.com/javamelody/javamelody/pull/965

evernat commented:
I think that this PR can't be merged, because:
The FilterServletResponseWrapper abstract class is extended by 3 classes in javamelody:
- CompressionServletResponseWrapper, which is used to compress the monitoring reports generated by javamelody. This filter response wrapper changes the length of the content with compression, that's why the setContentLength method was overwritten in this class (even if javamelody does not call this method).
- HtmlInjectorServletResponseWrapper, which is used to inject callback javascript into the webapp's html pages only when RUM is enabled via a javamelody parameter.
- CounterServletResponseWrapper, which is used always on all servlet responses of the webapp. This filter response wrapper only counts bytes and it does not change the length of the content. It should not touch anything on the reponse content or response headers. Doing otherwise would introduce behavior changes in the monitored webapp which must be avoided of course.

So do you agree that we should close this PR without merge?

evernat commented:
Or if the issue is the content length when RUM is enabled, you can create an issue to discuss that.

candrews commented:
Thank you for the review, explaining each class, and for pointing out those issues.

I've updated this PR in such a way that I believe addresses the feedback:
* I updated `CompressionServletResponseWrapper` so that it ignore any method which sets the Content-Length. I believe that is correct and intended.
* I updated `HtmlInjectorServletResponseWrapper` in the same way to also ignore any method which sets the Content-Length.
* Finally, I added another commit that creates a new class, `NoContentLengthFilterServletResponseWrapper`, which contains just the method overrides to ignore Content-Length setting, and have `CompressionServletResponseWrapper` and `HtmlInjectorServletResponseWrapper` extend it to eliminate duplication.

Notably, `FilterServletResponseWrapper` and `CounterServletResponseWrapper` are no longer touched.

This fixes the issue I was experiencing with RUM enabled where in my application, the HTML was truncated due to the Content-Length being shorter than the real body which includes the RUM injected script.

evernat commented:
> the issue I was experiencing with RUM enabled where in my application, the HTML was truncated due to the Content-Length being shorter than the real body which includes the RUM injected script.

Where and how have you seen that exactly? In a browser? Which one? What did you see?

candrews commented:
> Where and how have you seen that exactly? In a browser? Which one? What did you see?

You can reproduce this issue in a browser, but I find it easier to do so using curl. Here's how:

* Clone https://github.com/candrews/javamelody/tree/rum-issue - this is just JavaMelody with RUM enabled and the index page changed to include `</body`> so the RUM injector will do the injection
* Run the `javamelody-for-spring-boot` project
* Run `curl http://localhost:8080/ -H "accept: text/html"`

The actual output is:
```bash
$ curl http://localhost:8080/ -H "accept: text/html"  
<html><body>Greetings from Spring Boot!
<script src='
```

Note that the response is truncated. If you add the `-v` argument, you can see the header `Content-Length: 53` which explains where and why the truncation occurs.

The same behavior as curl exhibits can also be seen in browsers (I tested in Firefox and Chrome).

This PR fixes this truncation issue.

candrews commented:
@evernat is this now mergeable?

evernat commented:
I do not think that the CompressionServletResponseWrapper needs to be changed in this PR.
So, I would be ok to revert changes in the CompressionServletResponseWrapper  and to move methods from NoContentLengthFilterServletResponseWrapper to HtmlInjectorServletResponseWrapper.

candrews commented:
> I do not think that the CompressionServletResponseWrapper needs to be changed in this PR, at least for javamelody.
> So, I would be ok to revert changes in the CompressionServletResponseWrapper and to move methods from NoContentLengthFilterServletResponseWrapper to HtmlInjectorServletResponseWrapper.

Done! I've updates this PR accordingly.
